flax/io.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2022 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""IO Abstraction Layer.
+The sole purpose of this abstraction layer is to avoid requiring tensorflow
+as an open-source dependency solely for its tensorflow.io.gfile functions.
+"""
+import contextlib
+from enum import Enum
+import glob as glob_module
+import importlib
+import os
+import shutil
+
+from absl import logging
+from . import errors
+
+
+# Global Modes and selective import of tensorflow.io gfile.
+
+
+class BackendMode(Enum):
+  DEFAULT = 0
+  TF = 1
+
+io_mode = None
+gfile = None
+
+if importlib.util.find_spec('tensorflow'):
+  from tensorflow.io import gfile  # pytype: disable=import-error
+  io_mode = BackendMode.TF
+else:
+  logging.warning("Tensorflow library not found, tensorflow.io.gfile "
+                  "operations will use slower native shim calls.")
```

Comment:
Are native operations really slower than tensorflow?
Maybe worth pointing out that any access to GCS (i.e. "gs://" paths) will fail. Being explicit about this failure mode might also be a good idea inside below functions (e.g. when a program runs fine in one environment, but then crashes in another environment, it might not be obvious to the user that the 2nd invocation has crashed because it was trying to access a "gs://" path in an environment that did not have Tensorflow installed).

Reply:
I'm unsure on how the speed of native operations compare to tensorflow; maybe @levskaya can jump in.
As for GCS, yes good point, I'll add a disclaimer in the warning message.

Reply:
Created a new issue #2624